core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -0,0 +1,126 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.view;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Ordering;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract class ViewMetadata implements Serializable {
+  static final int DEFAULT_VIEW_FORMAT_VERSION = 1;
+  static final int SUPPORTED_VIEW_FORMAT_VERSION = 1;
+
+  public static ImmutableViewMetadata.Builder buildFrom(ViewMetadata metadata) {
+    return ImmutableViewMetadata.builder()
+        .formatVersion(DEFAULT_VIEW_FORMAT_VERSION)
+        .location(metadata.location())
+        .properties(metadata.properties())
+        .currentVersionId(metadata.currentVersionId())
+        .versions(metadata.versions())
+        .schemas(metadata.schemas())
+        .currentSchemaId(metadata.currentSchemaId())
+        .history(metadata.history());
+  }
+
+  @Value.Check
+  void check() {
+    Preconditions.checkState(versions().size() > 0, "Expecting non-empty version log");
+    checkVersionLog(history());
+  }
+
+  public abstract int formatVersion();
+
+  public abstract String location();
+
+  public abstract Map<String, String> properties();
+
+  public abstract int currentVersionId();
+
+  public abstract List<ViewVersion> versions();
+
+  public abstract List<ViewHistoryEntry> history();
+
+  @Nullable
+  public abstract List<Schema> schemas();
+
+  @Nullable
+  public abstract Integer currentSchemaId();
+
+  public ViewVersion version(int versionId) {
+    return versionsById().get(versionId);
+  }
+
+  @Value.Derived
+  public ViewVersion currentVersion() {
+    return versionsById().get(currentVersionId());
+  }
+
+  @Value.Derived
+  public Map<Integer, ViewVersion> versionsById() {
+    return indexVersions(versions());
+  }
+
+  @Value.Derived
+  public Map<Integer, Schema> schemasById() {
+    return indexSchemas(schemas());
+  }
+
+  @Value.Derived
+  public Schema schema() {
+    return schemasById().get(currentSchemaId());
+  }
+
+  private static void checkVersionLog(List<ViewHistoryEntry> versionLog) {
+    Preconditions.checkState(versionLog.size() > 0, "Expecting non-empty version log");
```

Comment:
```suggestion
    Preconditions.checkState(versionLog.size() > 0, "Expecting non-empty history log");
```

Reply:
otherwise the error msg might get confused for the one from L51

Reply:
I don't think this is a good idea. Why would we want to fail to read a view if the log is missing? I don't think that's a good reason to fail.

Reply:
If the log is missing this implies bad or corrupted view metadata. At minimum there must be at least 1 view version for the view to be considered as well defined. The original rationale was that if the metadata file corrupted, one shouldn't be able to read the view because there may be a correctness issue.

Alternatively, we could remove the validation since at the end of the day as long as current-version is well defined then the read can progress. With this approach I'm just concerned that if version-log is missing, and reads succeed then "corruption" just propagates silently as the view evolves and in the case of a genuine issue it can confuse someone trying to diagnose the issue.

 It seems better just to have a guarantee that version-log is well defined a the time of reading the view.

Reply:
After more thought, I think you're right @rdblue . For the purpose of validating view metadata, decouple the history of a view, which indicates when a view was considered "current" and the actual versions. We already verify that the current view version is valid at the time of reading the metadata which should be sufficient.